staging/src/k8s.io/apiserver/pkg/endpoints/filters/request_deadline_test.go

```diff
@@ -438,6 +443,961 @@ func TestWithFailedRequestAudit(t *testing.T) {
 	}
 }
 
+func TestPerHandlerWriteDeadlineHTTP2WithWriteShouldReturnErrorAfterTimeout(t *testing.T) {
+	// This test documents the behavior of the per handler write
+	// deadline with a standard net http2 server.
+	//
+	// scenario: http2, write deadline is set, after timeout occurs we expect
+	// the Write method of the ResponseWriter object to return an error.
+	t.Parallel()
+	clientDoneCh, handlerDoneCh := make(chan struct{}), make(chan error, 1)
+	server := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		defer close(handlerDoneCh)
+		if req.ProtoMajor != 2 {
+			t.Errorf("expected an HTTP/2.0 request, but got: %s", req.Proto)
+			return
+		}
+
+		ctrl := http.NewResponseController(w)
+		if err := ctrl.SetWriteDeadline(time.Now().Add(100 * time.Millisecond)); err != nil {
+			t.Errorf("expected no error from SetReadDeadline, but got: %v", err)
+			return
+		}
+
+		// a) the handler blocks here indefinitely
+		// b) this forces the handler write timeout
+		// to occur on the server side
+		// c) the http2 client should receive a stream error
+		// immediately after the write timeout occurs.
+		<-clientDoneCh
+
+		// NOTE: Write to the ResponseWriter object may succeed
+		// immediately after write timeout occurs, but we expect Write
+		// to return an error eventually; as soon as the underlying
+		// buffer is full, Write is expected to return an error.
+		func() {
+			now := time.Now()
+			count := 0
+			defer func() {
+				t.Logf("After timeout, Write was invoked %d times, duration: %s", count, time.Since(now))
+			}()
+
+			for {
+				count++
+				if _, err := w.Write(bytes.Repeat([]byte("a"), 1024)); err != nil {
+					handlerDoneCh <- err
+					break
+				}
+			}
+		}()
+	}))
+
+	defer server.Close()
+	server.EnableHTTP2 = true
+	server.StartTLS()
+
+	client := server.Client()
+	func() {
+		defer close(clientDoneCh)
+		// we intentionally don't set any transport or client timeout to
+		// simulate a default client-go client.
+		_, err := client.Get(server.URL + "/foo")
+		if !responsewritertesting.IsStreamReadOrWriteTimeout(err) {
+			t.Errorf("expected a stream reset error, but got: %v", err)
+		}
+	}()
+
+	select {
+	case err := <-handlerDoneCh:
+		if err == nil {
+			t.Errorf("expected an error from Write after timeout")
+			return
+		}
+		t.Logf("Write (invoked after timeout) returned an error: %v", err)
+	case <-time.After(wait.ForeverTestTimeout):
+		t.Errorf("expected the request handler to have terminated")
+	}
+}
+
+func TestPerHandlerWriteDeadlineHTTP1WithWriteShouldReturnErrorAfterTimeout(t *testing.T) {
```

Comment:
This function is very similar to TestPerHandlerWriteDeadlineHTTP2WithWriteShouldReturnErrorAfterTimeout ; I think it would be better if this were more DRY.